rpcserver.go

```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	}
+
+	var proofFile proof.File
+	if err := proofFile.Decode(bytes.NewReader(in.RawProof)); err != nil {
+		return nil, fmt.Errorf("unable to decode proof file: %w", err)
+	}
+
+	decodedProof, err := r.marshalProofFile(ctx, proofFile, in.ProofAtDepth,
+		in.WithPrevWitnesses, in.WithMetaReveal,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal proof: %w", err)
+	}
+
+	return &taprpc.DecodeProofResponse{
+		DecodedProof: decodedProof,
+	}, nil
+}
+
+// marshalProofFile turns a proof file into an RPC DecodedProof.
+func (r *rpcServer) marshalProofFile(ctx context.Context, proofFile proof.File,
+	index int64, withPrevWitnesses, withMetaReveal bool) (*taprpc.DecodedProof,
+	error) {
+
+	if index > int64(proofFile.NumProofs())-1 {
+		return nil, fmt.Errorf("invalid index %d is greater than latest proof "+
+			"index of %d", index, int64(proofFile.NumProofs())-1)
+	}
+
+	// Default to latest proof.
+	index = int64(proofFile.NumProofs()) - 1 - index
+
+	decodedProof, err := proofFile.ProofAt(uint32(index))
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		finalAsset     = decodedProof.Asset
+		rpcAsset       *taprpc.Asset
+		rpcMeta        *taprpc.AssetMeta
+		anchorOutpoint = wire.OutPoint{
+			Hash:  decodedProof.AnchorTx.TxHash(),
+			Index: decodedProof.InclusionProof.OutputIndex,
+		}
+		txMerkleProof  = decodedProof.TxMerkleProof
+		inclusionProof = decodedProof.InclusionProof
+		splitRootProof = decodedProof.SplitRootProof
+	)
+
+	var txMerkleProofBuf bytes.Buffer
+	if err := txMerkleProof.Encode(&txMerkleProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode serialized Bitcoin "+
+			"merkle proof: %w", err)
+	}
+
+	var inclusionProofBuf bytes.Buffer
+	if err := inclusionProof.Encode(&inclusionProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode inclusion proof: %w",
+			err)
+	}
+
+	var exclusionProofs [][]byte
+	for _, exclusionProof := range decodedProof.ExclusionProofs {
+		var exclusionProofBuf bytes.Buffer
+		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
+			return nil, fmt.Errorf("unable to encode exclusion "+
+				"proofs: %w", err)
+		}
+		exclusionProofBytes := exclusionProofBuf.Bytes()
+
+		exclusionProofs = append(exclusionProofs, exclusionProofBytes)
+	}
+
+	var splitRootProofBuf bytes.Buffer
+	if splitRootProof != nil {
+		if err := splitRootProof.Encode(&splitRootProofBuf); err != nil {
+			return nil, fmt.Errorf("unable to encode split root proof: %w",
+				err)
+		}
+	}
+
+	rpcAsset, err = r.marshalChainAsset(ctx, &tapdb.ChainAsset{
+		Asset:             &finalAsset,
+		AnchorTx:          &decodedProof.AnchorTx,
+		AnchorTxid:        decodedProof.AnchorTx.TxHash(),
+		AnchorBlockHash:   decodedProof.BlockHeader.BlockHash(),
+		AnchorOutpoint:    anchorOutpoint,
+		AnchorInternalKey: decodedProof.InclusionProof.InternalKey,
+	}, withPrevWitnesses)
+	if err != nil {
+		return nil, err
+	}
+
+	if withMetaReveal {
+		if len(rpcAsset.AssetGenesis.MetaHash) > 0 {
```

Comment:
I think we should also error out if we request meta reveal but don't have a meta hash. That would also remove one level of indentation, e.g.:
```
if withMetaReveal {
    if len(rpcAsset.AssetGenesis.MetaHash) == 0 {
        return nil, fmt.Errorf(....)
    }

    // fetch meta info ...
}
```